src/omotes_simulator_core/entities/assets/controller/profile_interpolation.py

```diff
@@ -0,0 +1,175 @@
+            logger.warning(
+                "No simulation timestep provided, using default of 3600 seconds (1 hour)"
+            )
+            timestep = 3600.0
+        else:
+            timestep = self.simulation_timestep
+
+        # Return empty profile if it has no data
+        if len(self.profile) < 2:
+            return self.profile
+
+        profile_timestep = float(
+            (
+                pd.to_datetime(self.profile["date"].iloc[1])
+                - pd.to_datetime(self.profile["date"].iloc[0])
+            ).total_seconds()
+        )
+
+        # Sample to greatest common divider - workaround the limitation of the rolling function
+        gcd_seconds = math.gcd(int(profile_timestep), int(timestep))
+        minor_step = pd.to_timedelta(gcd_seconds, unit="s")
+        index_minor = pd.date_range(
+            start=self.profile["date"].iloc[0], end=self.profile["date"].iloc[-1], freq=minor_step
+        )
+        profile_resample_minor = (
+            self.profile.set_index("date")
+            .reindex(index_minor)
+            .interpolate(method=self.interpolation_method.value)
+        )
+
+        points_in_window = int(timestep / gcd_seconds)
+        indexer = pd.api.indexers.FixedForwardWindowIndexer(window_size=points_in_window)
+
+        # Map sampling methods to their corresponding pandas rolling functions
+        rolling = profile_resample_minor["values"].rolling(window=indexer, min_periods=1)
+
+        profile_sampling_methods = {
+            ProfileSamplingMethod.ACTUAL: lambda: rolling.apply(lambda x: x.iloc[0]),
+            ProfileSamplingMethod.AVERAGE: lambda: rolling.mean(),
+            ProfileSamplingMethod.MAXIMUM: lambda: rolling.max(),
+            ProfileSamplingMethod.MINIMUM: lambda: rolling.min(),
+        }
+
+        if self.sampling_method not in profile_sampling_methods:
+            logger.error(f"Unknown sampling method: {self.sampling_method.value}")
+            raise ValueError(f"Unknown sampling method: {self.sampling_method.value}")
+
+        df_indexer = profile_sampling_methods[self.sampling_method]()
+        return pd.DataFrame({"date": df_indexer.index, "values": df_indexer.values})
+
+    def get_value(self, time: datetime.datetime) -> float:
+        """Get interpolated value at the specified time.
+
+        :param datetime.datetime time: Time for which to get the value
+        :return: Interpolated value
+        """
+        for index_resampled in range(self.start_index_resampled, len(self.resampled_profile)):
+            if (
+                abs(
+                    (
+                        self.resampled_profile["date"].iloc[index_resampled].to_pydatetime() - time
+                    ).total_seconds()
+                )
+                < 1e-6
+            ):
+                self.start_index_resampled = index_resampled
+                return float(self.resampled_profile["values"].iloc[index_resampled])
+        return 0.0
```

Comment:
It feels odd that you iterate over the complete data time list; you create a dataframe with a datetime index (resampled) can't you just use loc on that? 

This assumes that you have a correct index (e.g., starting at a defined t0) for your resampled dataframe. 

```suggestion
        try:
            return float(self.resampled_profile["values"].loc[index_resampled])
        except IndexError:
            return 0.0
```

Reply:
I agree.  I updated it based on the time, not the index_resampled.